### PR TITLE
Move expr_cast from refinement to util/convert_expr.h

### PR DESCRIPTION
**Pull request by romainbrenguier**
This moves these utility functions to util which is more appropriate.
The file and functions are also renamed to avoid confusion with util/expr_cast.h

### CODE REVIEW

File: src/util/convert_expr.h
Comment (tautschnig) [inline src/util/convert_expr.h]: Apparently this is more than just a plain file renaming? It's also a very confusing one, as there already are functions called `from_expr`.

**Review by tautschnig — DISMISSED** [review]
This is a very partial review: introducing `from_expr` with a very different meaning than the pre-existing `from_expr` API is confusing.

File: src/util/convert_expr.h
Comment (romainbrenguier) [reply src/util/convert_expr.h]: `from_expr` functions convert an expression to something else (which is also the case here). Here the type we are converting to is given as template argument so the meaning should be clear. But I'm open to suggestions for better names.

Comment (tautschnig) [reply src/util/convert_expr.h]: To be honest, I'm not very happy with the "convert" terminology. You are doing (partly) checked type casts, nothing more and nothing less. To me, "convert" suggests that there is some sort of transformation going on (which would be true of the existing `from_expr` code).

Comment (tautschnig) [reply src/util/convert_expr.h]: In that vein: `checked_cast_...` maybe? Also call the file `checked_expr_cast.h`?

Comment (romainbrenguier) [reply src/util/convert_expr.h]: Checked typecast is more what is done in expr_cast.h and it is a bit different from these functions.
In particular `expr_checked_cast` is already used there.

Comment (tautschnig) [reply src/util/convert_expr.h]: Honestly, I have no idea what this code is doing. The file used to be named "expr_cast.h", hence I made my inferences. There is a single comment in this file and having read that comment block three times now I still have no clue what it is trying to tell me (what is a "specific type", what is "empty" about an "optional" and how does an "optional" come into play when we are talking about `exprt`?). You are now saying it's "a bit different"? How much is "a bit"?

Comment (LAJW) [reply src/util/convert_expr.h]: (I wrote the original version). I wanted to follow the naming convention of `boost::lexical_cast`, only rather than throwing exceptions, I wanted to return empty optional to signal "bad conversion". It was developed independently of `expr_dynamic_cast`. But then the `expr_dynamic_cast` was merged (after `expr_cast` was already used locally in `string_refinement`). (`expr_dynamic_cast` and `expr_try_dynamic_cast` return references and pointers, but only of types derived from `exprt`).

Personally, I'd leave it as "expr_cast" and let it be the function which converts all sorts of `exprts` to various types in optional, and returning empty optional on error.

expr_cast.h doesn't actually have a function called "expr_cast", so it could be declared there.

Comment (tautschnig) [reply src/util/convert_expr.h]: You are making it sound like there are multiple somewhat competing implementations? That's interesting.

Maybe it would help (this might be what you are suggesting in your last sentence) to have all of this in one file? Then at least there would be more context to help understand the code.

Comment (LAJW) [reply src/util/convert_expr.h]: There is a single implementation for casting between `exprt`s with two different interfaces - throw on error or return `nullptr` on error (in `util/expr_cast.h`). This function however casts from `exprt` to an integer, string, or whatever other types we can reliably cast to. But since `exprt` is cheap to copy, it could also use the same pattern - `return optionalt<derived_exprt>`. Why it wasn't used in the first place - I have no idea.

Comment (tautschnig) [reply src/util/convert_expr.h]: Thank you very much for providing further detail -- what does it even mean to "cast" an `exprt` in this way? Anyway at least I believe I understand what the code is supposed to do. I have no idea how this code ever made it into the code base in its current form, but then that's a different question that the approvers of #1241 need to answer.

Anyway, it seems `util/arith_tools.h` seems to be the place for most of them.

Comment (LAJW) [reply src/util/convert_expr.h]: I don't think that conversion from `exprt` to string or from one `exprt` to another is an arithmetic operation.

It made its way into the code, as it provides saner way of doing those things compared to existing 15+ line dance with tons of `INVARIANTS` every time someone wanted to cast const_exprt to its C++ representation. The only "problem" is the name, which is debatable.

**Review by LAJW — APPROVED** [review]

File: src/util/convert_expr.h
Comment (tautschnig) [reply src/util/convert_expr.h]: > I don't think that conversion from exprt to string or from one exprt to another is an arithmetic operation.

Indeed, I generalised too quickly - the integer conversion ones should go in `arith_tools.h`, `exprt` to `*_exprt` should go in `expr_cast.h`, and `exprt` to string I have no idea what that is.

> The only "problem" is the name, which is debatable.

The problems certainly include the complete absence of documentation as well as poor choice of placement in the file system. Is it saner? I don't know, because I don't know what it's supposed to do (as there is no documentation).

Comment (smowton) [reply src/util/convert_expr.h]: Re: `expr_dynamic_cast` and returning pointers vs. `optionalt`, I believe @reuk was the person advocating pointers because the syntax to use them is briefer -- `optionalt<reference_wrapper<T>>` requires an explicit `.get()` operation a little too often.

**Review by smowton — CHANGES_REQUESTED** [review]
I think the two casts we can see here are different enough to deserve different names. The cast to `refined_string_exprt` should be a case of `expr_dynamic_cast` and implemented as such, whereas the to-integer conversions are helpers on top of `util/arith_tools.h`. I suggest:

1. Implement `can_cast_expr` and `validate_expr` in a string-refinement header (alongside the definition of `refined_string_exprt`) so that you can use `expr_dynamic_cast` with it.

2. Move the constant-expr-to-integer stuff into `arith_tools.h`. Deprecate the existing `to_integer` and `to_unsigned_integer` (but don't remove them) and replace with your new optional-returning or throwing versions.

File: src/util/convert_expr.h
Comment (LAJW) [reply src/util/convert_expr.h]: @smowton Before expr_dynamic_cast, there was this, which was intended to return `optional<T>` instead of `optional<reference_wrapper<T>>`. Old casts (`to_****_expr`) were doing that, because it's not inefficient, so I followed, as it was the safest and most concise way of doing those things.

**Comment by romainbrenguier** [comment]
@smowton Thanks for the suggestion. I didn't deprecate `to_unsigned_integer` yet because for now there is no `from_expr<unsigned>` implemented, but it would be easy to add.

**Comment by smowton** [comment]
Hmmm, two more nitpicks:

1. I think you should take `constant_exprt` rather than `exprt`, even though that's a bit more verbose at certain use-sites. It distinguishes failure because this was the wrong kind of exprt vs. failure because the constant wasn't numeric / was too large / whatever.

2. std::size_t's range may differ from `long` (as in `to_long()`), how about using `std::numeric_limits<...>.max()` etc to check in range, and thus generalise the `size_t` case into any integer type?

**Comment by tautschnig** [comment]
> std::size_t's range may differ from long (as in to_long()), how about using std::numeric_limits<...>.max() etc to check in range, and thus generalise the size_t case into any integer type?

There's `integer2size_t` covering this, and that's really what should be used here.

**Comment by smowton** [comment]
Whichever function does the conversion, I'd suggest templating on a numeric type rather than manually enumerating integer2long, integer2size_t, integer2uint16_t, ...

**Comment by tautschnig** [comment]
> Whichever function does the conversion, I'd suggest templating on a numeric type rather than manually enumerating integer2long, integer2size_t, integer2uint16_t, ...

True, but either way it should only be done in one place. If someone wants to step up with a PR doing the refactoring that's fine with me.

**Comment by romainbrenguier** [comment]
@smowton @tautschnig 
> Whichever function does the conversion, I'd suggest templating on a numeric type rather than manually enumerating integer2long, integer2size_t, integer2uint16_t, ...

Not sure that it is exactly what you had in mind but I implemented some conversion template which should help with that: https://github.com/diffblue/cbmc/pull/1582

**Comment by allredj** [comment]
Shall I review this or is it still likely to change?

**Comment by romainbrenguier** [comment]
@allredj please wait for https://github.com/diffblue/cbmc/pull/1582 to be merged, I will then adapt this one

**Comment by tautschnig** [comment]
It seems #1582 has been merged, so it may be possible to continue on this one?

**Comment by romainbrenguier** [comment]
@tautschnig yes but at the moment I'm blocked by the visual studio compilation problem. If someone has an idea?

**Comment by tautschnig** [comment]
I believe the VS failure is due to https://connect.microsoft.com/VisualStudio/feedback/details/1211985. You'll need to make those runtime checks for MSVC, I'm afraid.

**Comment by smowton** [comment]
Gah, https://connect.microsoft.com/VisualStudio/feedback/details/1211985

As @tautschnig says. https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B suggests `!defined(_MSC_VER) || _MSC_VER >= 1900` should diagnose a compiler with a properly `constexpr` `max()` method.

**Comment by romainbrenguier** [comment]
@tautschnig @smowton @allredj this is now working with VS and ready for rereview

File: src/util/mp_arith.h
Comment (tautschnig) [inline src/util/mp_arith.h]: s/with/will/? Also, the first sentence should say what it does, not just when it fails.

File: src/util/mp_arith.h
Comment (tautschnig) [inline src/util/mp_arith.h]: What does `_v` stand for?

File: src/util/string_expr.h
Comment (tautschnig) [inline src/util/string_expr.h]: Use `validate_operands`?

File: src/solvers/refinement/string_constraint_generator_transformation.cpp
Comment (tautschnig) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: Why does this one work without a second template argument?

File: src/solvers/refinement/string_refinement.cpp
Comment (tautschnig) [inline src/solvers/refinement/string_refinement.cpp]: The counterpart of the above comment: why does this require a second template argument?

**Review by tautschnig — COMMENTED** [review]
I'm confused by the sometimes-use-a-second-template-argument, and if the second template argument (which always is `exprt`) is truly required then this is not at all beautiful.

**Comment by smowton** [comment]
Given that the possible source types are exactly `exprt` and `constant_exprt`, would it be simplest to omit the source template parameter and just introduce overloads for the two cases, one of which hard-casts the `exprt` to `constant_exprt` and calls the other?

**Comment by romainbrenguier** [comment]
@smowton 
> Given that the possible source types are exactly exprt and constant_exprt, would it be simplest to omit the source template parameter and just introduce overloads for the two cases, one of which hard-casts the exprt to constant_exprt and calls the other?

Not sure I understood this correctly, but `mp_integer` is also a possible source type.

**Comment by smowton** [comment]
As in you can `numeric_cast` from `mp_integert` to `size_t` for example? Can you also `numeric_cast` between primitives?

**Comment by romainbrenguier** [comment]
@smowton 
> As in you can numeric_cast from mp_integert to size_t for example? 

Yes you can (and to any integral type)

> Can you also numeric_cast between primitives?

Not at the moment but it could be added if someone finds it useful.

**Comment by smowton** [comment]
OK, understood. In this case removing the source template arg is beyond my skill level, but @reuk or @LAJW can probably work magic if asked nicely :)

**Comment by reuk** [comment]
I'm pretty sure that's possible, yeah. I think you just need an overload of `numeric_cast` in the struct for each possible `Source` type.

**Comment by reuk** [comment]
Having said that, this does require moving the `exprt` overload back into the declaration in mp_arith.h, which I think is what you're trying to avoid...

**Comment by reuk** [comment]
Created a [pr showing how this might look](https://github.com/romainbrenguier/cbmc/pull/1)

File: src/util/string_expr.h
Comment (allredj) [inline src/util/string_expr.h:247]: Shouldn't we check anything about the type of operands?

File: src/util/arith_tools.h
Comment (allredj) [inline src/util/arith_tools.h]: Could you please add the same comment as below?

File: src/util/arith_tools.h
Comment (allredj) [inline src/util/arith_tools.h]: This seems to be the same function as above, except for the `mpi` function called. Could we refactor?

File: src/util/arith_tools.h
Comment (allredj) [inline src/util/arith_tools.h]: Why do we call that `Target` here and not T? Update the doc header.

File: src/util/arith_tools.h
Comment (allredj) [inline src/util/arith_tools.h]: `anexpression` -> `an expression`

File: src/util/arith_tools.h
Comment (allredj) [inline src/util/arith_tools.h:133]: What does `_v` mean?

File: src/util/arith_tools.h
Comment (romainbrenguier) [reply src/util/arith_tools.h]: I refactored a bit to have only one structure for the conversion to integral type, but I still have one function for mpi to signed and one function from mpi to unsigned because all the checks are different.

File: src/util/arith_tools.h
Comment (romainbrenguier) [reply src/util/arith_tools.h]: this is to make it clear this is the type we convert to, and not from.

File: src/util/arith_tools.h
Comment (romainbrenguier) [reply src/util/arith_tools.h:133]: It stands for value (this is to distinguish from optionalt). This was the notation used in the `expr_cast` I removed, but I'm open to better suggestions.

File: src/util/arith_tools.h
Comment (reuk) [reply src/util/arith_tools.h]: Are you sure? I'm pretty sure the checks can be made the same (I show how this is done in my proposal PR)

Comment (romainbrenguier) [reply src/util/arith_tools.h]: Yes it can be done, I will try that.

File: src/util/arith_tools.h
Comment (allredj) [reply src/util/arith_tools.h:133]: `_val`?

**Review by smowton — APPROVED** [review]

File: src/solvers/refinement/string_constraint_generator_transformation.cpp
Comment (tautschnig) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: `std::size_t` for consistency. (Also applies above.)

File: src/solvers/refinement/string_refinement.cpp
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement.cpp]: Why do you use sometimes unsigned and sometimes sizet for values that denote sizes?

File: src/solvers/refinement/string_refinement.cpp
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement.cpp:828]: I'd appreciate braces here.

**Review by peterschrammel — APPROVED** [review]
Lgtm

**Review by allredj — APPROVED** [review]

**Pull request merged**
